pubtools/_content_gateway/push_base.py

```diff
@@ -0,0 +1,463 @@
+            1) The key must be of tuple type
+            2) Three keys must be present in the key tuple
+
+        Args:
+            key (tuple(int, int, int)):
+                The keys should contain product, product_code and product_version
+
+        Raises:
+            ValueError:
+                When the keys are not of tuple type or
+                the keys in tuple doesn't satisfy the length size.
+        """
+
+        if not isinstance(key, tuple):
+            raise ValueError("key must be tuple (product, product_code, product_version)")
+        elif not len(key) == 3:
+            raise ValueError("key must be tuple (product, product_code, product_version)")
+
+    @staticmethod
+    def _file_key_check(key):
+        """
+        Checks the product version mapping keys.
+        The keys must follow the following constraints:
+            1) The key must be of tuple type
+            2) Four keys must be present in the key tuple
+
+        Args:
+            key (tuple(int, int, int)):
+                The keys should contain product, product_code,
+                 product_version and downloadURL
+
+        Raises:
+            ValueError:
+                When the keys are not of tuple type or
+                the keys in tuple doesn't satisfy the length size.
+        """
+
+        if not isinstance(key, tuple):
+            raise ValueError("key must be tuple (product, product_code, product_version, downloadURL)")
+        elif not len(key) == 4:
+            raise ValueError("key must be tuple (product, product_code, product_version, downloadURL)")
+
+    def _fetch_product(self, product_name, product_code):
+        """
+        Get the list of products from the content gateway
+        by calling its client methods.
+
+        Args:
+            product_name (str):
+                product name of the product
+            product_code (str):
+                product code of the product
+
+        Returns:
+            Yields (dict):
+                Returns the product record. The returned structure
+                is an iterator to reduce memory requirements.
+
+                Intended to be called as the first operation in
+                the content gateway workflow.
+        """
+
+        all_products = self.cgw_client.get_products()
+        for product in all_products:
+            yield (product['name'], product['productCode']), product['id']
+
+    def _fetch_product_version(self, product_name, product_code, version_name):
+        """
+        Get the list of product versions from the content gateway
+        by calling its client methods.
+
+        Args:
+            product_name (str):
+                product name of the product
+            product_code (str):
+                product code of the product
+            version_name (str):
+                version name of the product
+
+        Returns:
+            Yields (dict):
+                Returns the version records associate with the product.
```

Comment:
Only some version records are returned? And some extra data is returned as well?